### ISSUE TITLE
Mutation's "inherited" state

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/tskit#2631]
Getting the state that a mutation inherited is fiddly right now, and has been annoying me quite bit recently. We should add a method to the Python object at least to get it, but it would also be nice to have access to it in numpy for reasoning efficiently about reversions and so on.

Does ``inherited_state`` feel like the right term here? 

The numpy stuff would depend on us making some decisions about vectorised access to the string columns which I guess I should open a seperate issue on.

```python
tables = ts.tables
assert np.all(tables.mutations.derived_state_offset == np.arange(ts.num_mutations + 1))
derived_state = tables.mutations.derived_state.view("S1")
assert np.all(tables.sites.ancestral_state_offset == np.arange(ts.num_sites + 1))
ancestral_state = tables.sites.ancestral_state.view("S1")
# Compute the "inherited state" for each mutation
inherited_state = ancestral_state[ts.mutations_site]
mutations_with_parent = ts.mutations_parent != -1
parent = ts.mutations_parent[mutations_with_parent]
assert np.all(parent >= 0)
inherited_state[mutations_with_parent] = derived_state[parent]

# Check
for mut in ts.mutations():
    state0 = ts.site(mut.site).ancestral_state
    if mut.parent != -1:
        state0 = ts.mutation(mut.parent).derived_state
    assert state0 == inherited_state[mut.id].decode()
```

**Comment by benjeffery** [comment]
`inherited_state` sounds good to me, only other thing I could come up with `parent_state` but I think that could get confused with the parent node state.

**Comment by hyanwong** [comment]
I have wanted this a number of times too, so +1 from me.

**Comment by jeromekelleher** [comment]
Yes, it's a bug really that we don't fully validate the mutation parents on tree sequence creation.

**Comment by benjeffery** [comment]
Closing for inactivity and labelling "future", please re-open if you plan to work on this.

**Comment by hyanwong** [comment]
I definitely want to implement this once the mutation parents bug has been fixed, so I'm assigning it to me, and reopening. I'll at least implement the non-array version.

**Comment by jeromekelleher** [comment]
Yep - and their inherited state is the site's ancestral state. This is really for complicated stuff like SARS-CoV-2 where we get up to 30K mutations per site.

**Comment by molpopgen** [comment]
Oh -- I get it, it is the state inherited by the node where a mutation first lands.

**Comment by hyanwong** [comment]
Now that https://github.com/tskit-dev/tskit/pull/3212#event-19538220537 and https://github.com/tskit-dev/tskit/pull/3228 are merged, we can finally add this, I think. It would be *really* useful.

I think we should probably have `ts.mutation(N).inherited_state()` too, should we?
